tests/test_dense.py

```diff
@@ -0,0 +1,69 @@
+from qutip_cupy import CuPyDense
+import numpy as np
+import pytest
+
+@pytest.fixture(scope="function", params=((1, 2), (5, 10), (7, 3), (2, 5)))
+def shape(request):
+    return request.param
+
+
+class TestCuPyDenseDispatch:
+    """ Tests if the methods and conversions have been
+        succesfully registered to QuTiP's Data Layer."""
+
+    def test_conversion_cycle(self, shape):
+
+        from qutip.core import data
+
+        qutip_dense = data.Dense(np.random.uniform(size=shape))
+
+        tr1 = data.to(CuPyDense, qutip_dense)
+        tr2 = data.to(data.Dense, tr1)
+
+        np.testing.assert_array_equal(qutip_dense.to_array(), tr2.to_array())
+
+class TestCuPyDense:
+    """ Tests of the methods and constructors of the CuPyDense class. """
+
+    def test_shape(self, shape):
+
+        cupy_dense = CuPyDense(np.random.uniform(size=shape))
+
+        assert (cupy_dense.shape == shape)
+
+    def test_transpose(self, shape):
+
+        cupy_dense = CuPyDense(np.random.uniform(size=shape)).transpose()
+        
+        np.testing.assert_array_equal(cupy_dense.shape, (shape[1],shape[0]))
```

Comment:
```suggestion
        data = np.random.uniform(size=shape)
        cd = CuPyDense(data).transpose()
        assert cd.shape == (shape[1], shape[0])
        np.testing.assert_array_equal(cd.to_array(), data.transpose())
```
We also need to assert that the transposed dense array has the correct numbers in it!

Reply:
np.testing.assert_array_equal  as per description checks the equality of the shapes already.

Reply:
unless we want a separate check on the shape to be sure it it not a failure in copying the values.

Reply:
I'm happy for us to delete the assert on the shape.